File: src/testsuites/vec4.c
#include <stdio.h>
#include <cu/cu.h>
#include <fermat/vec4.h>
#include <fermat/dbg.h>
#include "data.h"

TEST(vec4SetUp)
{
}

TEST(vec4TearDown)
{
}

TEST(vec4Alloc)
{
    fer_vec4_t *v, w;

    v = ferVec4New(0., 1., 2., 3.);
    ferVec4Set(&w, 0., 1., 2., 3.);
    assertTrue(ferVec4Eq(v, &w));
    ferVec4Del(v);
}

TEST(vec4Add)
{
    fer_vec4_t v;
    size_t i;

    printf("# ---- add ----\n");
    ferVec4Set(&v, FER_ZERO, FER_ZERO, FER_ZERO, FER_ZERO);
    for (i = 0; i < vecs4_len; i++){
        ferVec4Add(&v, &vecs4[i]);
        printf("# %g %g %g %g\n", ferVec4X(&v), ferVec4Y(&v), ferVec4Z(&v), ferVec4W(&v));
    }
    printf("# ---- add end ----\n\n");
}

TEST(vec4Sub)
{
    fer_vec4_t v;
    size_t i;

    printf("# ---- sub ----\n");
    ferVec4Set(&v, FER_ZERO, FER_ZERO, FER_ZERO, FER_ZERO);
    for (i = 0; i < vecs4_len; i++){
        ferVec4Sub(&v, &vecs4[i]);
        printf("# %g %g %g %g\n", ferVec4X(&v), ferVec4Y(&v), ferVec4Z(&v), ferVec4W(&v));
    }
    printf("# ---- sub end ----\n\n");
}

TEST(vec4Scale)
{
    fer_vec4_t v;
    size_t i;

    printf("# ---- scale ----\n");
    for (i = 0; i < vecs4_len; i++){
        ferVec4Copy(&v, &vecs4[i]);
        ferVec4Scale(&v, ferVec4X(&vecs4[i]));
        printf("# %g %g %g %g\n", ferVec4X(&v), ferVec4Y(&v), ferVec4Z(&v), ferVec4W(&v));
    }
    printf("# ---- scale end ----\n\n");
}

TEST(vec4Normalize)
{
    fer_vec4_t v;
    size_t i;

    printf("# ---- normalize ----\n");
    for (i = 0; i < vecs4_len; i++){
        ferVec4Copy(&v, &vecs4[i]);
        ferVec4Normalize(&v);
        printf("# %g %g %g %g\n", ferVec4X(&v), ferVec4Y(&v), ferVec4Z(&v), ferVec4W(&v));
    }
    printf("# ---- normalize end ----\n\n");
}

TEST(vec4Dot)
{
    fer_real_t dot;
    size_t i;

    printf("# ---- dot ----\n");
    for (i = 0; i < vecs4_len - 1; i++){
        dot = ferVec4Dot(&vecs4[i], &vecs4[i + 1]);
        printf("# %g\n", dot);
    }
    printf("# ---- dot end ----\n\n");
}

TEST(vec4Mul)
{
    fer_vec4_t v;
    size_t i;

    printf("# ---- mul ----\n");
    for (i = 0; i < vecs4_len - 1; i++){
        ferVec4MulComp2(&v, &vecs4[i], &vecs4[i + 1]);
        printf("# %g %g %g %g\n", ferVec4X(&v), ferVec4Y(&v), ferVec4Z(&v), ferVec4W(&v));
    }
    printf("# ---- mul end ----\n\n");
}


TEST(vec4Len2)
{
    fer_real_t d;
    size_t i;

    printf("# ---- len2 ----\n");
    for (i = 0; i < vecs4_len; i++){
        d = ferVec4Len2(&vecs4[i]);
        printf("# %g\n", d);
    }
    printf("# ---- len2 end ----\n\n");
}

TEST(vec4Len)
{
    fer_real_t d;
    size_t i;

    printf("# ---- len ----\n");
    for (i = 0; i < vecs4_len; i++){
        d = ferVec4Len(&vecs4[i]);
        printf("# %g\n", d);
    }
    printf("# ---- len end ----\n\n");
}

TEST(vec4Dist2)
{
    fer_real_t d;
    size_t i;

    printf("# ---- dist2 ----\n");
    for (i = 0; i < vecs4_len - 1; i++){
        d = ferVec4Dist2(&vecs4[i], &vecs4[i + 1]);
        printf("# %g\n", d);
    }
    printf("# ---- dist2 end ----\n\n");
}

TEST(vec4Dist)
{
    fer_real_t d;
    size_t i;

    printf("# ---- dist ----\n");
    for (i = 0; i < vecs4_len - 1; i++){
        d = ferVec4Dist(&vecs4[i], &vecs4[i + 1]);
        printf("# %g\n", d);
    }
    printf("# ---- dist end ----\n\n");
}

File: python/testsuites/vec4.py
import unittest
from common import *
from fermat import *


class TestCase(TC):
    def __init__(self, mn):
        TC.__init__(self, mn, 'Vec4')

    def testNumberProtocol(self):
        self.out('---- NumberProtocol ----')
        a = Vec4()
        self.out(a)
        a = Vec4(1)
        self.out(a)
        a = Vec4(1, 2)
        self.out(a)
        a = Vec4(1, 2, 3)
        self.out(a)
        a = Vec4(1, 2, 3, 4)
        self.out(a)

        self.out()

        b = Vec4(3.1, 1.2, 0.4, 11.2)
        self.out(str(a), '+', str(b), '=', str(a + b))
        self.out(str(a), '+', 1., '=', str(a + 1.))
        self.out(1., '+', str(a), '=', str(1. + a))

        self.out(str(a), '-', str(b), '=', str(a - b))
        self.out(str(a), '-', 1., '=', str(a - 1.))

        self.out(str(a), '*', str(b), '=', str(a * b))
        self.out(str(a), '*', 1.5, '=', str(a * 1.5))
        self.out(1.5, '*', str(a), '=', str(1.5 * a))

        self.out(str(-a))

        self.out(str(a), '/', 2, '=', str(a / 2))

        self.out()
        b = Vec4(3.1, 1.2, 0.4)
        b += a
        self.out(str(b))
        b += 10.
        self.out(str(b))
        b -= a
        self.out(str(b))
        b -= 10
        self.out(str(b))
        b *= 2.
        self.out(str(b))
        b /= 2.
        self.out(str(b))

    def testSeqProtocol(self):
        self.out('---- SeqProtocol ----')
        a = Vec4(1, 2, 3, 4)
        self.out(len(a), a[0], a[1], a[2], a[3])
        self.out(a.x, a.y, a.z, a.w)
        for i in a:
            self.out(i)

    def testCmp(self):
        self.assertTrue(Vec4(1, 2, 3, 4) == Vec4(1, 2, 3, 4))
        self.assertTrue(Vec4(1, 2, 3, 4) != Vec4(2, 2, 3, 4))

    def testMethods(self):
        self.out('---- Methods ----')
        a = Vec4(1, 2, 3)
        b = a.copy()
        b += 1
        self.out(a, b)

        self.out('len', a.len())
        self.out('len2', a.len2())
        self.out('dist', a.dist(b))
        self.out('dist2', a.dist2(b))
        b.scaleToLen(4.)
        self.out('scaleToLen', b, b.len())
        self.out('scaledToLen', b.scaledToLen(2.), b.len(), b.scaledToLen(2.).len())
        self.out('normalized', b.normalized(), b.len(), b.normalized().len())
        b.normalize()
        self.out('normalize', b, b.len())
